listmatch/listmatch.py

```diff
@@ -12,27 +12,26 @@ def __init__(self):
 
 class StateSet(object):
     def __init__(self, states):
-        self.states = set()
-        for state in states:
-            self.add_state(state)
+        self.states = {}
+        for state, paths in states:
+            self.add_state(state, paths)
 
-    def add_state(self, state):
+    def add_state(self, state, paths):
         if state in self.states:
+            self.states[state] += paths
             return
-        self.states.add(state)
+        self.states[state] = paths
         for eps in state.epsilon:
-            self.add_state(eps)
+            self.add_state(eps, paths)
 
     def __iter__(self):
-        return iter(self.states)
+        return iter(self.states.items())
```

Comment:
ping @pokey  do you care about preserving order here? if so, Python dict doesn't guarantee that, but if you do care about it you can use [OrderedDict](https://docs.python.org/2/library/collections.html#collections.OrderedDict) which is available in Python 2.7 and up